### PR TITLE
Support a deny list

### PR DESCRIPTION
**Pull request by gotjosh**
With this, we can block certain emails from ever reaching our email
provider.

### CODE REVIEW

**Review by dannykopping — CHANGES_REQUESTED** [review]
LGTM overall, minor questions

File: main.go [hunk 1]
```diff
@@ -100,27 +100,36 @@ func senderChecker(peer smtpd.Peer, addr string) error {
...
+				Warn("denied_recipients invalid")
+			return observeErr(smtpd.Error{Code: 451, Message: "Invalid recipient address"})
+		}
 
-	log.WithField("address", addr).
-		Warn("Invalid recipient address")
-	return observeErr(smtpd.Error{Code: 451, Message: "Invalid recipient address"})
+		if allowedRegexp.MatchString(addr) && !deniedRegexp.MatchString(addr) {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply main.go]: Good shout.

File: main.go [hunk 2]
```diff
@@ -100,27 +100,36 @@ func senderChecker(peer smtpd.Peer, addr string) error {
 	return observeErr(smtpd.Error{Code: 451, Message: "sender address not allowed"})
 }
 
-func recipientChecker(peer smtpd.Peer, addr string) error {
-	if *allowedRecipients == "" {
-		// allow any recipient, disable recipient check
-		return nil
-	}
+func recipientChecker(allowed, denied string) func(peer smtpd.Peer, addr string) error {
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply main.go:103]: This makes it easier to test and more explicit, I think this is the natural evolution of a codebase and not necessarily an inconsistency. 

It's the first function we actually test, others might follow suit as they deemed fit.

File: main_test.go [hunk 3]
```diff
@@ -23,17 +23,32 @@ func Test_RecepientsCheck(t *testing.T) {
 		denied   string
 		expected error
 	}{
+		{
+			name:   "with emails that are denied",
+			emails: []string{"delivery@example.com", "example@email.com", "<example@email.com>"},
```
Comment (dannykopping) [inline main_test.go:28]: How will these be denied if there's no denyList defined?

File: main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dannykopping) [reply main.go:103]: Fair enough :+1: works for me (didn't make the connection to the tests, that sways me)

**Review by dannykopping — APPROVED** [review]
Nice one @gotjosh, thanks for the changes
:shipit:

File: main_test.go (hunk 3 above)
Comment (gotjosh) [reply main_test.go:28]: The wording was off, thanks!

**Pull request merged**